### PR TITLE
Docker helper script for CHT 4.x

### PR DESCRIPTION
**Pull request by m5r**
# Description

Adds one liner support for app developers to easily start up a new CHT instance(s).

See [nascent docs](https://docs.communityhealthtoolkit.org/apps/guides/hosting/4.x/app-developer/#cht-docker-helper-for-4x).

medic/cht-core#7852

To test:
* check out this branch
* cd into `scripts/docker-helper-4.x`
* run `./cht-docker-compose.sh` OR `./cht-docker-compose.sh EXISTING-PROJECT.env` OR `./cht-docker-compose.sh EXISTING-PROJECT.env down` OR  `./cht-docker-compose.sh EXISTING-PROJECT.env destroy`
* data is stored in `~/.medic/`

# Code review checklist

- [X] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [X] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [X] Tested: Unit and/or e2e where appropriate

# Compose URLs:

* CHT Core: CHT_CORE_COMPOSE_URL
* CouchDB Single: COUCH_SINGLE_COMPOSE_URL
* CouchDB Cluster: COUCH_CLUSTER_COMPOSE_URL
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by mrjones-plip** [comment]
Really awesome to see this coming together!  First try it just worked - TLS cert and all - this is sweeeet! CouchDB data is in [its own directory](https://github.com/medic/cht-core/pull/7871/files#diff-9b27c58372bc9326462c3e882509fe6b70afa1227c01598ea9a501e4eb3acdf7R45) - lovely! 

I know this is still in a draft, but there's some early gotchas I wanted to point out in my testing:
* we don't catch port conflicts, so we'll [endlessly print out](https://github.com/medic/cht-core/pull/7871/files#diff-9b27c58372bc9326462c3e882509fe6b70afa1227c01598ea9a501e4eb3acdf7R76) "waiting for the container.." because the `nginx` container can never start. Maybe give up after 10 tries? Or show a port conflict? Or both? End result should be actionable by user.
* On that same note - we [hard code the ports](https://github.com/medic/cht-core/pull/7871/files#diff-9b27c58372bc9326462c3e882509fe6b70afa1227c01598ea9a501e4eb3acdf7R38) to always be `8080` and `8443` - this means an app dev will never be able to run two concurrent instances without editing the env files. Maybe prompt which ports with a default of `8080` and `8443`
* we should show  the URL with IP and port (eg `https://192-168-68-26.my.local-ip.co:8443`) upon successful start up. NB - We already do this by first [getting the host IP](https://github.com/medic/cht-core/blob/master/scripts/docker-helper/cht-docker-compose.sh#L24), then [building up](https://github.com/medic/cht-core/blob/master/scripts/docker-helper/cht-docker-compose.sh#L45) the *.my.local-ip.co URL.
* we want this to (hopefully!) run on macOS, Windows and Linux -  `openssl` isn't a [reasonable expectation](https://github.com/medic/cht-core/pull/7871/files#diff-9b27c58372bc9326462c3e882509fe6b70afa1227c01598ea9a501e4eb3acdf7R44) .  since we don't need hardcore random-ness of `openssl` - [maybe something simpler](https://stackoverflow.com/questions/1194882/how-to-generate-random-number-in-bash)? 
* I'm not entirely sure - but I think per [this comment](https://github.com/medic/cht-core/issues/7852#issuecomment-1285127347) - we need to save off separate core and couch compose files per instance vs [the shared copies that are in place now](https://github.com/medic/cht-core/issues/7852#issuecomment-1285127347). This is so the upgrader service can edit the compose files and it won't affect the other instances using the same files.
* user input is hard! I messed up the system by entering a project name of `deleteme 2`.  It created a file called `deletem 2.env`, but then I had two choices to start project `deleteme` or project `2`. On my next attempt I put in a filename of `../../foo` and it, unsurprisingly, created a file two directories up, but then showed no projects available to start.

**Comment by mrjones-plip** [comment]
Checking some more, I think we're instantiating the upgrade service wrong some how.  
*  I believe we're supposed to have one upgrade service container per CHT instance. Right now this script does just one shared, global instance
* when i run `docker logs compose-files_cht-upgrade-service_1` I see errors like the ones below:
   ```
   Encountered errors while bringing up the project.

   Error while upgrading Error: The COUCHDB_UUID variable is not set. Defaulting to a blank string.
   The DOCKER_CONFIG_PATH variable is not set. Defaulting to a blank string.
   The CHT_COMPOSE_PATH variable is not set. Defaulting to a blank string.
   ```
* I'm unable to upgrade instances, say to `7778-bulk-delete-redesign (~4.0.0)` branch - which should work.

**Comment by mrjones-plip** [comment]
One more update:
* if you chose`N` when prompted `Would you like to initialize a new project [y/N]?` , and you have no existing .env files, you're stuck in an infinite loop of `Which project do you want to use?`

**Comment by mrjones-plip** [comment]
@m5r  - I tried to work through all the outstanding issues to get this closer to an MVP release. I think we're very close! I crossed them out as I fixed them and only the "can't upgrade" one is left undone (see below).

Please feel free to take this PR over or let it sit until I get back on Nov 14th.  

**Status**: Everything works, including multiple concurrent versions, but for some reason upgrade the cht-upgrade-service container dies.  Pretty sure all my fixes introduced a bug somewhere! :mag:

**Comment by mrjones-plip** [comment]
**Update**!  - All known issues resolved. Please test/comment/merge/get another review from Binod - what ever makes sense!

**Comment by lorerod** [comment]
Thanks,  @m5r and @mrjones-plip, for this script and, [this](https://www.youtube.com/watch?v=-x9zr8R2hP4) detail video on how to use it.
I followed the video, and I'm getting a different outcome. The script ends with the `Upgrade service no longer running - check for port conflicts or other errors and try again.` message. However, the upgrade service continued logging until the `4_oh_first` instance was successfully running.
<details>
<summary>Docker helper script log:</summary>

```
marialorenarodriguezviruel@MacBook-Pro-de-Maria docker-helper-4.x % ./cht-docker-compose.sh                                             
No project found, follow the prompts to create a project .env file.
Would you like to initialize a new project [y/N]? y
How do you want to name the project? 4-oh-first
Which project do you want to use?
1) 4_oh_first
#? 1

Creating network "4_oh_first-cht-net" with the default driver
Pulling cht-upgrade-service (public.ecr.aws/s5s3h4s7/cht-upgrade-service:latest)...
latest: Pulling from s5s3h4s7/cht-upgrade-service
9621f1afde84: Pull complete
17cb27c841b2: Pull complete
a9cbc71018b1: Pull complete
2377ef6be762: Pull complete
fcef55c4d6e6: Pull complete
c07d03ffb4d3: Pull complete
7705c6b431d7: Pull complete
Digest: sha256:f8042c5878156144f2e2105a5135a399bf7e32e67521dcf930cbdc17a7a5ade0
Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-upgrade-service:latest
Creating 4_oh_first-dir_cht-upgrade-service_1 ... done
Starting project "4_oh_first". First run takes a while. Will try for up to five minutes...

Upgrade service no longer running - check for port conflicts or other errors and try again.

marialorenarodriguezviruel@MacBook-Pro-de-Maria docker-helper-4.x % 
```

</details>

<details>
<summary>Upgrade service log:</summary>

```
Listening on port 5008

networks:

  cht-net:

    name: 4_oh_first-cht-net

services:

  api:

    depends_on:

      haproxy:

        condition: service_started

    environment:

      BUILDS_URL: https://staging.dev.medicmobile.org/_couch/builds_4

      COUCH_URL: http://medic:password@haproxy:5984/medic

      UPGRADE_SERVICE_URL: http://cht-upgrade-service:5008

    expose:

    - '5988'

    image: public.ecr.aws/s5s3h4s7/cht-api:4.0.0-alpha

    logging:

      driver: local

      options:

        max-file: '20'

        max-size: 50m

    networks:

      cht-net: null

    restart: always

  haproxy:

    environment:

      COUCHDB_PASSWORD: ****

      COUCHDB_SERVERS: couchdb

      COUCHDB_USER: medic

      HAPROXY_IP: haproxy

      HAPROXY_PORT: '5984'

      HEALTHCHECK_ADDR: healthcheck

    expose:

    - '5984'

    hostname: haproxy

    image: public.ecr.aws/s5s3h4s7/cht-haproxy:4.0.0-alpha

    logging:

      driver: local

      options:

        max-file: '20'

        max-size: 50m

    networks:

      cht-net: null

    restart: always

  healthcheck:

    environment:

      COUCHDB_PASSWORD: ****

      COUCHDB_SERVERS: couchdb

      COUCHDB_USER: medic

    image: public.ecr.aws/s5s3h4s7/cht-haproxy-healthcheck:4.0.0-alpha

    logging:

      driver: local

      options:

        max-file: '20'

        max-size: 50m

    networks:

      cht-net: null

    restart: always

  nginx:

    depends_on:

      api:

        condition: service_started

      haproxy:

        condition: service_started

    environment:

      API_HOST: api

      API_PORT: '5988'

      CERTIFICATE_MODE: SELF_SIGNED

      COMMON_NAME: test-nginx.dev.medicmobile.org

      COUNTRY: US

      DEPARTMENT: Information_Security

      EMAIL: domains@medic.org

      LOCALITY: San_Francisco

      ORGANISATION: medic

      SSL_CERT_FILE_PATH: /etc/nginx/private/cert.pem

      SSL_KEY_FILE_PATH: /etc/nginx/private/key.pem

      STATE: California

    image: public.ecr.aws/s5s3h4s7/cht-nginx:4.0.0-alpha

    logging:

      driver: local

      options:

        max-file: '20'

        max-size: 50m

    networks:

      cht-net: null

    ports:

    - published: 8080

      target: 80

    - published: 8443

      target: 443

    restart: always

    volumes:

    - cht-ssl:/etc/nginx/private:rw

  sentinel:

    depends_on:

      haproxy:

        condition: service_started

    environment:

      API_HOST: api

      COUCH_URL: http://****:password@haproxy:5984/medic

    image: public.ecr.aws/s5s3h4s7/cht-sentinel:4.0.0-alpha

    logging:

      driver: local

      options:

        max-file: '20'

        max-size: 50m

    networks:

      cht-net: null

    restart: always

version: '3.9'

volumes:

  cht-ssl: {}



networks:

  cht-net:

    name: 4_oh_first-cht-net

services:

  couchdb:

    environment:

      COUCHDB_LOG_LEVEL: error

      COUCHDB_PASSWORD: ****

      COUCHDB_SECRET: fa94810e3d8bde010b01f7be41538df2

      COUCHDB_USER: medic

      COUCHDB_UUID: 723aca3d4d176f5f5033c5c93f018375

      SVC_NAME: couchdb

    image: public.ecr.aws/s5s3h4s7/cht-couchdb:4.0.0-alpha

    logging:

      driver: local

      options:

        max-file: '20'

        max-size: 50m

    networks:

      cht-net: null

    restart: always

    volumes:

    - /Users/marialorenarodriguezviruel/.medic/cht-docker/4_oh_first-dir/couch:/opt/couchdb/data:rw

    - cht-credentials:/opt/couchdb/etc/local.d:rw

version: '3.9'

volumes:

  cht-credentials: {}



networks:

  cht-net:

    name: 4_oh_first-cht-net

services:

  cht-upgrade-service:

    environment:

      BUILDS_SERVER: null

      CERTIFICATE_MODE: null

      CHT_COMPOSE_PATH: /Users/marialorenarodriguezviruel/.medic/cht-docker/4_oh_first-dir

      CHT_COMPOSE_PROJECT_NAME: 4_oh_first

      CHT_NETWORK: 4_oh_first-cht-net

      COMMON_NAME: null

      COUCHDB_DATA: /Users/marialorenarodriguezviruel/.medic/cht-docker/4_oh_first-dir/couch

.IsADirectoryError: [Errno 21] Is a directory: '/docker-compose/couch'


      COUCHDB_PASSWORD: ****

      COUCHDB_SECRET: fa94810e3d8bde010b01f7be41538df2

      COUCHDB_SERVERS: null

      COUCHDB_USER: medic

      COUCHDB_UUID: 723aca3d4d176f5f5033c5c93f018375

      COUNTRY: null

      DEPARTMENT: null

      DOCKER_CONFIG: /config

      DOCKER_CONFIG_PATH: /Users/marialorenarodriguezviruel/.medic/cht-docker/4_oh_first-dir

      EMAIL: null

      LOCALITY: null

      MARKET_URL_READ: null

      NGINX_HTTPS_PORT: '8443'

      NGINX_HTTP_PORT: '8080'

      ORGANISATION: null

      SSL_CERT_FILE_PATH: null

      SSL_KEY_FILE_PATH: null

      SSL_VOLUME_MOUNT_PATH: null

      STATE: null

    image: public.ecr.aws/s5s3h4s7/cht-upgrade-service:latest

    networks:

      cht-net: null

    volumes:

    - /var/run/docker.sock:/var/run/docker.sock:rw

    - /Users/marialorenarodriguezviruel/.medic/cht-docker/4_oh_first-dir:/config:rw

    - /Users/marialorenarodriguezviruel/.medic/cht-docker/4_oh_first-dir:/docker-compose:rw

version: '3.9'



4.0.0-alpha: Pulling from s5s3h4s7/cht-couchdb


Digest: sha256:380612edd8f6762fe7a07068d0eb8a9281e27d2994ac4aa3a2a98465903237f2


Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-couchdb:4.0.0-alpha


4.0.0-alpha: Pulling from s5s3h4s7/cht-haproxy


Digest: sha256:7a4beb544e9c03a4522d7a01e8aab77ef3b5a4071b49ff874bb57693610b3619


Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-haproxy:4.0.0-alpha


4.0.0-alpha: Pulling from s5s3h4s7/cht-sentinel


Digest: sha256:f51af6dc67fa252a08d01246de735b1bbda031e65e0c56ffeb8668d52ef35bab


Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-sentinel:4.0.0-alpha


4.0.0-alpha: Pulling from s5s3h4s7/cht-api


Creating volume "4_oh_first_cht-ssl" with default driver


Creating volume "4_oh_first_cht-credentials" with default driver


Pulling couchdb (public.ecr.aws/s5s3h4s7/cht-couchdb:4.0.0-alpha)...


Pulling haproxy (public.ecr.aws/s5s3h4s7/cht-haproxy:4.0.0-alpha)...


Pulling sentinel (public.ecr.aws/s5s3h4s7/cht-sentinel:4.0.0-alpha)...


Pulling api (public.ecr.aws/s5s3h4s7/cht-api:4.0.0-alpha)...


Digest: sha256:a7a9c3f976fcb9cc5b92f0cca108ff33de08862df5613ff0f7fb053c754473f8


Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-api:4.0.0-alpha


4.0.0-alpha: Pulling from s5s3h4s7/cht-nginx


Digest: sha256:6530ba2493025e1b3327d03f061c64ef8dd498281e20ebc2deb4010e63ffcc8d


Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-nginx:4.0.0-alpha


4.0.0-alpha: Pulling from s5s3h4s7/cht-haproxy-healthcheck


Digest: sha256:358608c03b161fe270df9cc685b8b844dfc6460da557b87afd0278dcc5e6aa16


Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-haproxy-healthcheck:4.0.0-alpha


Pulling nginx (public.ecr.aws/s5s3h4s7/cht-nginx:4.0.0-alpha)...


Pulling healthcheck (public.ecr.aws/s5s3h4s7/cht-haproxy-healthcheck:4.0.0-alpha)...


Creating 4_oh_first_healthcheck_1 ... 


Creating 4_oh_first_haproxy_1     ... 


Creating 4_oh_first_cht-upgrade-service_1 ... 


Creating 4_oh_first_couchdb_1             ... 


Creating 4_oh_first_healthcheck_1         ... done


Creating 4_oh_first_cht-upgrade-service_1 ... done


Creating 4_oh_first_haproxy_1             ... done


Creating 4_oh_first_api_1                 ... 


Creating 4_oh_first_sentinel_1            ... 


Creating 4_oh_first_couchdb_1             ... done


Creating 4_oh_first_sentinel_1            ... done


Creating 4_oh_first_api_1                 ... done


Creating 4_oh_first_nginx_1               ... 


Creating 4_oh_first_nginx_1               ... done
```

</details>

<details>
<summary>My local instance is up and running with 4.0.0-alpha:</summary>

Captura de Pantalla 2022-11-11 a la(s) 12 40 48


</details>

Please let me know if you need more information.

**Comment by mrjones-plip** [comment]
@lorerod  - Thanks for the test and early feedback! Always helpful to get detailed input like this.

Could you try removing the docker containers and storage volumes and run `git fetch && git pull origin` (or the Docker Desktop equivalent) and try again? There was a bug I found and I fixed near the end of my development (see 7ea32baa7a4953d9da2e3c1779045ff894f64001). It addresses the specific upgrade service crashing which you're experiencing.  Alternately, you can check if you're on that hash with this call, running it in the `cht-core` repo:  `git log --oneline --decorate | head -n 1`

The reason is that if you hadn't update your `cht-core` repo before testing the script it would manifest this bug. I'm thinking you might be on on older revision.

If the issue persists, can you try rerunning the command and give me the name of your project (eg `4_oh_first`) and the output of this command: `docker ps --format '{{.Names}}'` . I ask because we're checking for a specific container name (eg `4_oh_first-cht-upgrade-service`) and we may have it wrong!

**NB** -  when the upgrade service crashes the cht core/couchdb services will still run - this is why it worked in your browser!

**Comment by lorerod** [comment]
Hi @mrjones-plip thank you for the detail explanation. 

- I ran `git fetch && git pull origin`

- The output of `git log --oneline --decorate | head -n 1` is `7ca168f2b (HEAD -> 7852-cht-4.x-docker-helper, origin/7852-cht-4.x-docker-helper) add faucton URL`

- I named the project `4-oh-third`
- The output of `docker ps --format '{{.Names}}'` is:
```
4_oh_third_nginx_1
4_oh_third_api_1
4_oh_third_sentinel_1
4_oh_third_haproxy_1
4_oh_third_cht-upgrade-service_1
4_oh_third_healthcheck_1
4_oh_third_couchdb_1
4_oh_third-dir_cht-upgrade-service_1
```

<details>
<summary>cht-docker-compose.sh log:</summary>

```
% ./cht-docker-compose.sh                   
No project found, follow the prompts to create a project .env file.
Would you like to initialize a new project [y/N]? y
How do you want to name the project? 4-oh-third
Which project do you want to use?
1) 4_oh_third
#? 1

Creating network "4_oh_third-cht-net" with the default driver
Pulling cht-upgrade-service (public.ecr.aws/s5s3h4s7/cht-upgrade-service:latest)...
latest: Pulling from s5s3h4s7/cht-upgrade-service
9621f1afde84: Pull complete
17cb27c841b2: Pull complete
a9cbc71018b1: Pull complete
2377ef6be762: Pull complete
fcef55c4d6e6: Pull complete
c07d03ffb4d3: Pull complete
7705c6b431d7: Pull complete
Digest: sha256:f8042c5878156144f2e2105a5135a399bf7e32e67521dcf930cbdc17a7a5ade0
Status: Downloaded newer image for public.ecr.aws/s5s3h4s7/cht-upgrade-service:latest
Creating 4_oh_third-dir_cht-upgrade-service_1 ... done
Starting project "4_oh_third". First run takes a while. Will try for up to five minutes...

Upgrade service no longer running - check for port conflicts or other errors and try again.
```
</details>

**Comment by mrjones-plip** [comment]
Thanks for the feedback @lorerod ! I'll dig into your questions and get back to you.

**Comment by mrjones-plip** [comment]
@m5r - I see two current issues with the script:

1. Some how `nginx` can become unhappy an endlessly restart with a TLS cert error. I've had trouble reproducing:
   ```
   Launching Nginx
   CERTIFICATE MODE = SELF_SIGNED
   /docker-entrypoint.sh: Configuration complete; ready for start up
   2022/11/15 05:37:16 [emerg] 1#1: SSL_CTX_use_PrivateKey("/etc/nginx/private/key.pem") failed (SSL: error:0B080074:x509 certificate routines:X509_check_private_key:key values mismatch)
   nginx: [emerg] SSL_CTX_use_PrivateKey("/etc/nginx/private/key.pem") failed (SSL: error:0B080074:x509 certificate routines:X509_check_private_key:key values mismatch)
   ```
1. An update to the script I made is causing two update containers :upside_down_face: 
   ```
   docker ps --format '{{.Names}}'|grep upgr
   one_more_time_please_cht-upgrade-service_1
   one_more_time_please-dir-cht-upgrade-service-1
   ```

**Comment by m5r** [comment]
@mrjones-plip how can I reproduce both of them?  
From reading the code, it should always use the one with `-dir` but I'm trying to replicate the bug locally

**Comment by mrjones-plip** [comment]
@lorerod - after talking to @m5r today and reviewing feedback from other teammates, we found a few bugs with the script.  One of them is the name of the upgrade service (eg `4_oh_third_cht-upgrade-service_1` vs `4_oh_third-dir_cht-upgrade-service_1`).  Your feedback was really helpful in uncovering this and verifying what we, and other teammates on Slack, were seeing!  Thanks very much.

When we have something more to show for our efforts, we'll report back in the Slack thread.

Cheers!

**Comment by mrjones-plip** [comment]
@m5r = I was thinking on the multiple upgrade containers and then it hit me: the upgrade service runs ALL compose files in the `CHT_COMPOSE_PATH` dir - including itself!  Splitting off the core and couchdb into a subdir `compose` solved the problem.

For the time being I've removed the upgrade service check from the while loop as well.  we can add it back if we want later.

I've updated the [last comment](https://github.com/medic/cht-core/pull/7871#issuecomment-1314802073) with the to-dos

**Comment by tatilepizs** [comment]
Thank you again @m5r for this helpful script.

I have one question, last week I used it to test another branch, I installed it and everything was working fine, and at the end of my day I stopped the containers. Today I ran the script again to continue with my testing, I choose the option to use the project that I created last week, but when the containers were up again it was installed version `master (~4.0.0)` it reverted the branch that I installed.

I don't know if this is something that you already considered, but I wanted to ask if it will be possible that when I start a project that was already created it remains in the branch that was installed?

**Comment by mrjones-plip** [comment]
@tatilepizs  - Excellent discover!  I have confirmed that when you stop and start the containers, the CHT does revert to `4.0.0-alpha-*` version.

This is because we download the compose [files every time](https://github.com/medic/cht-core/blob/7852-cht-4.x-docker-helper/scripts/docker-helper-4.x/cht-docker-compose.sh#L96). instead of doing it only when [we first initialize the files](https://github.com/medic/cht-core/blob/7852-cht-4.x-docker-helper/scripts/docker-helper-4.x/cht-docker-compose.sh#L70)  An easy fix!  

cc @m5r 

(PS - I'm not actually working this week, but this piqued my curiosity so I took a look-see!)

**Comment by mrjones-plip** [comment]
@jkuester offered to grab this - thanks for letting me volunteer you though @garethbowen  ;)

**Review by jkuester — COMMENTED** [review]
Very nice!  This runs super fast for me!  I am not a bash expert, but I have been line-by-line through the script and things look fine to me.  (Excellent use of the `docker --filter`!  :heart:) 
 Just a couple of comments of varying importance:

- [ ] 1. There is no usage documentation!  I can never remember the exact syntax of commands and so rely heavily on printing docs using `--help`!
    - It is nice that it will walk me through spinning up a new project if I just run `./cht-docker-compose.sh` but my first question, after I successfully got it to start was "how do I stop it?"
    - Might be nice to include the commands for `down` and `destroy` as a part of the summary that is printed after the instance starts...
- [ ] 2. The `Which project do you want to use? ` menu is great! Unfortunately there does not seem to be any way to use it with the `down` or `destroy` commands. To make matters worse, if you use an invalid `.env` name with one of those commands, the `Which project do you want to use? ` menu is triggered but behaves unexpectedly!  Example Scenario:
    - I create a project called `my_first_4`
    - I stop my project by calling `./cht-docker-compose.sh my_first_4.env down`
    - I try to destroy my project by calling `./cht-docker-compose.sh my_first_4 destroy` (missing the `.env` file extension. I made this mistake pretty much every time I tried to call down or destroy while I was testing it....)
    - I see:

```
File my_first_4 doesnt exist
Would you like to initialize a new project [y/N]? 
Which project do you want to use? (ctrl + c to quit) 
1) my_first_4
```

-
    - I enter `1`
    - Instead of destroying `my_first_4`, the script just re-starts it.  It seems like in this case it would be best to not display the `Which project do you want to use? ` menu at all.
- [ ] 3. Alternatively, it would be really nice to have a way to use the `Which project do you want to use? ` menu for all the commands so that I do not have to remember the exact name of all my projects (and I don't have to remember to include the `.env` prefix at the end of the name when running the command).  I feel like this could be a feature we add later, though. Not needed for the MVP.
- [ ] 4. I was surprised that the data/compose files were automatically stored in `.medic` but the auto-created `.env` files were stored in `docker-helper-4.x/.`.  It might be less confusing if these were both stored at the same place (and would perhaps help avoid weirdness when I have multiple copies of the cht-core repo checked out and end up running the `cht-docker-compose.sh` from various places at different times....
    - My vote would be to store everything in `.medic` 
- [ ] 5. Can we consider changing the `down` command to just be `stop` instead?  I know this wording is coming from the original script, but now seems like a good time to lobby for it to be changed!  :smile: In the Docker compose world, `down` means "stop and remove" all the containers (so basically it does exactly what our `destroy` command does).

**Comment by mrjones-plip** [comment]
>  Just a couple of comments of varying importance:

Thanks for the detailed responses @jkuester! I'll dig though them in the coming day or so and get back to you.  Your fresh set of eyes expose some weaknesses that will be great to improve.

**Comment by mrjones-plip** [comment]
Thanks again for all the thought and input on this! Really appreciate it. 

Enumerated item feedback above:

1. awesome feedback! added help at intro and exit in fef94d97f.  Try calling it now with `--help` or `-h` !
2. Oh...yeah - interesting. I ALWAYS just used tab completion so never encountered this.  Good catch!  fixed in 33716065e.
3. Ah - yeah - ah more unified approach would be nice.  Skipping for MVP though. That said, I got really used using tab complete to start stuff quickly
4. I'm gonna push back on this one. I think very few app devs will have multiple copies of cht-core around.  Furhter, the script works autonomously, and the docs call out how to run it with `curl` outside of cht-core repo.  I'll see if my push back sticks - open to chat if you feel this still fundementally not ok!
5. I recently switched it from "nuke" -> "down", so a change is easy! Further, your logic of using "stop" is sound. Done in 2567d54e9

**Comment by mrjones-plip** [comment]
back to you sir @jkuester for more another review ( :crossed_fingers: build passes too!!)

File: scripts/docker-helper-4.x/cht-docker-compose.sh [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    show_help_intro
+    show_help_existing_stop_and_destroy
+    exit 0
+	elif [[ -f "$1" ]]; then
+		projectFile=$1
+		projectName=$(echo "$projectFile" | sed "s/\.\///" | sed "s/\.env//")
+		homeDir=$(get_home_dir "$projectName")
+	else
+	  echo ""
+		echo -e "${red}File \"$1\" doesnt exist - be sure to include \".env\" at the end!${noColor}"
```
Comment (jkuester) [inline scripts/docker-helper-4.x/cht-docker-compose.sh:102]: Thanks for this!  :laughing:

**Review by jkuester — APPROVED** [review]
:star: :star: :star: :star: :star:

So great!

Regarding `#4`, this is absolutely not a sticking point for me!  (Happy to approve as is.)  I see the benefit of having the .env files local to the script (tab-complete FTW!). The thing that I got hung up on here was that at one point I manually deleted the folder for a project from `.medic` and then was surprised when the script told me I could not create another project with that name because the project still existed...  But, if you use the `destroy` command properly, you should not have this problem!

**Comment by mrjones-plip** [comment]
Thanks so much @jkuester ! The PR stars are in alignment - after you lgtm - the CI passed!!!

**Pull request merged**